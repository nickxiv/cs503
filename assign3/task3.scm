
(define (deque)
    (define front nil)
    (define back nil)
    (define size 0)

    (define (enqueueFront val)
        (define temp (node val nil nil))
        (cond 
            ((nil? front)           ;empty queue 
                (set! back temp)
            )
            (else
                ((front'set-prev!) temp)
                ((temp'set-next!) front)
            )
        )
        (set! front temp)
        (set! size (+ size 1))
    )
    (define (enqueueBack val)
        (define temp (node val nil nil))
        (cond 
            ((nil? back)           ;empty queue 
                (set! front temp)
            )
            (else
                ((back'set-next!) temp)
                ((temp'set-prev!) back)
                
            )
        )
        (set! back temp)
        (set! size (+ size 1))

    )

    (define (enqueueIndex index val)
        (cond 
            ((= 0 index) (enqueueFront val))
            ((= size index) (enqueueBack val))
            (else 
                (define (helper source nodeBefore)
                    (cond 
                        ((= index source) 
                            (define temp (node val (nodeBefore'next) nodeBefore))
                            (cond 
                                ((not (nil? (nodeBefore'next))) 
                                    (((nodeBefore'next)'set-prev!) temp)
                                )
                            )
                            ((nodeBefore'set-next!) temp)
                            (set! size (+ size 1))
                        )
                        (else 
                            (helper (+ source 1) (nodeBefore'next))
                        )
                    )
                )
                (helper 1 front)
            )
        )
    )

    (define (dequeueIndex index)
        (cond
            ((= 0 index) (dequeueFront))
            ((= (- size 1) index) (dequeueBack))
            (else 
                (define (helper source curr)
                    (cond 
                        ((= index source)
                            (((curr'next)'set-prev!) (curr'prev))
                            (((curr'prev)'set-next!) (curr'next))
                            (set! size (- size 1))
                            (curr'value)
                        )
                        (else
                            (helper (- source 1) (curr'prev))
                        )
                    )
                )
                (helper (- size 1) back)
            )
        )
    )

    (define (dequeueFront)
        (cond 
            ((= size 0) (println "Dequeue from empty list"))
            (else
                (define returnVal (front'value))
                (cond 
                    ((not (nil? (front'next))) (((front'next)'set-prev!) nil))
                )
                (set! front (front'next))
                (set! size (- size 1))
                returnVal
            )
        )
    )

    (define (dequeueBack)
    (cond 
        ((= size 0) (println "Dequeue from empty list"))
        (else
            (define returnVal (back'value))
            (cond
                ((not (nil? (back'prev))) (((back'prev)'set-next!) nil))
            )
            (set! back (back'prev))

            (set! size (- size 1))
            returnVal
            )
    )
)


    (define (display)
        (define (helper source)
            (cond 
                ((nil? source) (print "]"))
                (else 
                    (print (source'value))
                    (if (not (nil? (source'next))) (print ","))
                    (helper (source'next))
                )
            )
        )
        (print "[")
        (helper front)
    )
    this
)

(define (node value next prev)
    (define (set-next! val)
        (set! next val)
    )
    (define (set-prev! val)
        (set! prev val)
    )
    (define (set-value! val)
        (set! value val)
    )
    this
)
(define (main)
    (setPort (open (getElement ScamArgs 1) 'read))
    (define env this)
    (define (iter expr)
        (if (not (eof?)) 
            (begin (eval expr env) (iter (readExpr)))
        )
    )
    (iter (readExpr))    
)