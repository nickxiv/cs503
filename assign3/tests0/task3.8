(include "random.lib")
(randomSeed 106)
(define dq (deque))
(define (upback dq n)
    (define value (randomRange 10 100))
    (cond
        ((= n 0)
            ((dq'enqueueBack) value)
            )
        (else
            ((dq'enqueueBack) value)
            (upback dq (- n 1))
            )
        )
    )
(define (upfront dq n)
    (define value (randomRange 10 100))
    (cond
        ((= n 0)
            ((dq'enqueueFront) value)
            )
        ((< (dq'size) 10)
            ((dq'enqueueFront) value)
            (upfront dq (- n 1))
            )
        (else
            ((dq'enqueueIndex) (randomRange 2 10) value)
            (upfront dq (- n 1))
            )
        )
    )
(define (downback dq n)
    (cond
        ((or (= n 0) (< (dq'size) 10))
            ((dq'dequeueIndex) 0)
            )
        (else
            ((dq'dequeueIndex) (- (dq'size) (randomRange 2 10)))
            (downback dq (- n 1))
            )
        )
    )
(upback dq 2541)
(upfront dq 2598)
(downback dq 4950)
((dq'display)) (println)
