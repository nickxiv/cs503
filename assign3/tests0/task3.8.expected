[97,78,30,89,38,83,54,77,23,57,94,95,55,27,70,70,53,19,30,13,28,97,11,57,70,40,91,70,64,54,77,39,15,22,88,49,29,24,80,63,39,51,59,28,25,48,79,48,96,39,81,95,69,14,77,88,81,65,10,15,59,25,40,60,45,40,97,61,27,55,71,37,19,45,33,47,79,79,26,27,56,37,99,19,29,35,86,95,64,24,53,35,75,94,42,84,82,14,88,28,23,24,95,79,80,56,84,15,50,26,41,91,61,12,12,67,23,53,61,58,92,64,49,71,55,12,63,79,48,30,85,25,51,33,16,40,71,30,41,46,45,67,68,37,79,25,82,99,92,63,14,21,99,19,81,72,73,39,49,74,17,60,82,23,23,22,98,46,87,99,40,51,48,93,32,97,36,15,33,56,24,13,24,72,66,84,55,57,21,75]
